backend/src/resolvers.js

```diff
@@ -0,0 +1,73 @@
+const { UserInputError } = require("apollo-server");
+
+const resolvers = {
+    Query: {
+        users: (parent, args, context) => context.dataSources.db.allUsers(),
+        posts: (parent, args, context) => context.dataSources.db.allPosts(),
+    },
+    Mutation: {
+        write: async (parent, args, context) => {
+            const newPost = {
+                title: args.post.title,
+                authorName: args.post.author.name,
+            };
+            const author = await context.dataSources.db.getUser(newPost.authorName);
+            if (!author) {
+                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
+            }
+            return await context.dataSources.db.createPost(newPost);
+        },
+        delete: async (parent, args, context) => {
+            const id = args.id;
+            const post = await context.dataSources.db.getPost(id);
+            if (!post) {
+                throw new UserInputError("Invalid post", { invalidArgs: id });
+            }
+            return await context.dataSources.db.deletePost(id);
+        },
+        upvote: async (parent, args, context) => {
+            const name = args.voter.name;
+            const id = args.id;
+            const upvoter = await context.dataSources.db.getUser(name);
+            if (!upvoter) {
+                throw new UserInputError("Invalid user", { invalidArgs: name });
+            }
+            const post = await context.dataSources.db.getPost(id);
+            if (!post) {
+                throw new UserInputError("Invalid post", { invalidArgs: id });
+            }
+            return await context.dataSources.db.upvotePost(id, name);
+        },
+        downvote: async (parent, args, context) => {
+            const name = args.voter.name;
+            const id = args.id;
+            const downvoter = await context.dataSources.db.getUser(name);
+            if (!downvoter) {
+                throw new UserInputError("Invalid user", { invalidArgs: name });
+            }
+            const post = await context.dataSources.db.getPost(id);
+            if (!post) {
+                throw new UserInputError("Invalid post", { invalidArgs: id });
+            }
+            return await context.dataSources.db.downvotePost(id, name);
+        },
+    },
+    Post: {
+        author: async (obj, args, context) => {
+            return await context.dataSources.db.getUser(obj.authorName);
+        },
+        votes: async (obj, args, context) => {
+            let values = Array.from(obj.voters.values());
+            let votes = values.reduce((sum, number) => sum + number, 0);
+            return votes;
```

Comment:
This is not going to be an issue for our miniature application, but calculating the number of votes on each request seems way to inefficient compared to how simple it is to calculate it in advance.

I don't know Apollo good enough to tell whether it can do some caching here, but even if it does I would try to circumvent caching if possible, because, you know
>There are only two hard things in Computer Science: cache invalidation and naming things.

Reply:
You *could* setup a cache for this but then it's yet another cache (a piece of state) that you are responsible to keep up to date. My suggestion: Aim for absence of errors (e.g. avoid state) and solve performance problems when you have them. 

We developers have a tendency for premature optimization. Try to look for common performance problems (N+1, sequential execution, network), most often your application code is not the defining performance bottleneck. 

Also: Calculating the `votes` on the resolver has the performance advantage, that you only calculate it when requested:
```gql
{
  posts {
    title
    votes
  }
}
```
Will call `Post.votes` resolver, but:
```gql
{
  posts {
    title
  }
}
```
will not.